kythe/go/extractors/config/preprocessor/README.md

```diff
@@ -0,0 +1,73 @@
+# Build Config Preprocessing
+
+A number of builders require custom modification to a repo's build config before
+Kythe extraction will work correctly.
+
+## Preprocessor
+
+This is the top level binary for doing preprocessing.  The default operation is
+simply `./preprocess <build-file>`, which will detect what format the build file
+is in and perform necessary steps.
+
+## Invoking via docker / Cloud Build
+
+TODO(#3158): Actually fill this out with a Dockerfile and example commands.
+
+## Specific Builders
+
+### Gradle
+
+In order to run Kythe extraction on a gradle repo, we must first modify the
+`gradle.build` file to hook into a separate javac wrapper binary.
+`gradle_build_modifier.go` takes an input `gradle.build` file and appends the
+bits necessary for replacing javac calls with Kythe's `javac-wrapper.sh`.
+
+```
+allprojects {
+  gradle.projectsEvaluated {
+    tasks.withType(JavaCompile) {
+      options.fork = true
+      options.forkOptions.executable = '/opt/kythe/extractors/javac-wrapper.sh'
+    }
+  }
+}
+```
+
+If the input file already contains reference to
+`options.forkOptions.executable`, then `gradle_build_modifier.go` does nothing.
+
+#### Future work
+
+The current implementation uses simple string-based matching, without actually
+understanding the structure.  If that becomes necessary in the future, it might
+be better to use the existing Java libraries for `org.codehaus.groovy.ast` to
+properly parse the build.gradle file and have more precise picture.  In
+particular `org.codehaus.groovy.ast.CodeVisitorSupport` might be sufficient.
+Ideally we find such a library in golang.
+
+### Maven
+
+Maven's build config is handled by `pom_xml_modifier.go`.  It utilizes the etree
+xml library to parse and modify the mvn `pom.xml` config file in a similar way
+as described above for gradle.  One notable difference between gradle and maven
+here is that gradle actually embeds the refrence to the javac wrapper directly
+into the build file, while the modifications to the maven pom xml file merely
+allow future configuration at runtime.
+
+The example snippet it drops in is:
+
+```
+<build>
+  <plugins>
+    <plugin>
+      <groupId>org.apache.maven.plugins</groupId>
+      <artifactId>maven-compiler-plugin</artifactId>
+      <version>3.7.0</version>
+      <configuration>
+        <source>1.8</source>
+        <target>1.8</target>
```

Comment:
What about other Java versions?

Reply:
TBD still #3075

In particular my first efforts towards getting this to work have failed.